webrender/src/render_task.rs

```diff
@@ -252,6 +253,20 @@ impl RenderTask {
 
             match clip_info.bounds.inner {
                 Some(ref inner) if !inner.device_rect.is_empty() => {
+                    // If this clip item has an inner rect that completely
+                    // encloses the screen rect of the primitive, it can't
+                    // possibly affect the result, so remove it during
+                    // this filtering pass.
+                    if inner.device_rect.contains_rect(&prim_rect) {
```

Comment:
There is only one case where `task_rect` is larger than `prim_rect`: when we are processing clips specified for parent stacking contexts, while the primitive itself doesn't have a specific clip.

But in this case it should be desirable to have this task even if the particular primitive rect is not affected, because the generated mask is to be re-used for multiple primitives inside this stacking context.

I think we need to sort this out at a higher level: have a heuristic for a stacking context to share it's mask with children primitives. The heuristic can be as follows:
  1. compute the bounding box of all primitives inside this stacking context that don't have specific clips already
  2. compute the sum of bounding box squares of all the same primitives
  3. if the sum is smaller than the bounding box of all, then pass `None` here for `clip_id` and `prim_rect` for `task_rect`
  4. if the sum is larger, then do what we are doing now

As you can see, with this approach we'll not need a hack inside `new_mask`, and we'll not pass the primitive rectangle in here (just the task rectangle).

Reply:
That does sound a bit more elegant conceptually, but I think the implementation may be a bit complicated.

We'd have to either scan the primitive list inside a stacking context first to get the sum, and know which path to choose when generating the masks, or we'd have to keep track of the primitives that need a clip mask and then revisit them to set up the appropriate clip mask, wouldn't we? It's possible I'm over-thinking this :)

I think we could land this first, since it's a simple win, and then revisit the heuristic above as a possible follow up. What do you think?

Reply:
Huh, I have an idea that might fix this and drastically simplify the way we manage shared masks. Will need some more thought and prototyping to see if it works though. @jrmuizel Do you want an urgent fix for that performance issue, or is it fine to hold off for a few days and work out the best long term solution?